### DISCUSSION TITLE
Add option to pause based on other process on system being active

### DISCUSSION THREAD

**Discussion by opentoe** [sabnzbd/sabnzbd#2232] [Ideas]
I know there is a schedule in sab where you can turn it off/on at certain times which is great, but I would love something smarter. I would love an option that tells SABNZBD to STOP/CEASE all functions when a movie/media starts playing and then will continue where it left off. This would be fantastic. Sometimes SABNZBD is doing some heavy downloading and extracting 30GB+ files, and several times I have had trouble with the video playback. Are there any devs brave enough to tackle something this difficult? Of course it has to be fully automated. Finish watching a movie and your DL queue just gets larger. I think this would be a refined option, plex pass only. Thanks for reading and let me know what you know!

**Comment by thezoggy** [comment]
sab has no awareness of what else is running on the box nor do we care to. sab does have an api, you could make a script to pause/resume sab via it. or you could just move sab to another box or use any of the os level options or even sab built in options to control resources such as only process top item / pause during post processing / use nice

**Comment by Safihre** [comment]
Actually, this is maybe a tiny bit interesting actually. Because with Pyton we can query the active processes on the system and take an action based on them.

**Comment by sanderjo** [comment]
I could use it too. So while VLC is playing content, I don't want SAB to do something, because that will cause stalling in VLC.

I thought it was easy: check CPU en MEM of vlc:

```
$ ps -aux | grep -e USER -e vlc
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
sander    163335  3.2 76.5 9156396 6043544 ?     Sl   jun29 198:31 /usr/bin/vlc --started-from-file /media/zeegat/blabla.mkv
```

But ... no matter if vlc is playing or paused, CPU and MEM are at the same level.

It seems on my system, disk io is the bottleneck. Which I can monitor with iotop (sudo needed), and maybe with iostat (not yet clear how).

Bonus problem: in iotop, vlc disk usage will jump up and down between 0 and high (600 KB/s) during usage.

EDIT:

```
$ python3
Python 3.8.10 (default, Mar 15 2022, 12:22:08) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import psutil
>>> p = psutil.Process(163335)
>>> p.name()
'vlc'
>>> p.exe()
'/usr/bin/vlc'
>>> p.io
p.io_counters(  p.ionice(       
>>> p.io_counters()
pio(read_count=827611, write_count=5612474, read_bytes=1917616128, write_bytes=12288, read_chars=815908407, write_chars=373782879)
>>> p.io_counters()
pio(read_count=828692, write_count=5613516, read_bytes=1918533632, write_bytes=12288, read_chars=816720673, write_chars=373807010)
>>> p.io_counters()
pio(read_count=828993, write_count=5613805, read_bytes=1918533632, write_bytes=12288, read_chars=816721684, write_chars=373814161)
>>> p.io_counters()
pio(read_count=830528, write_count=5615294, read_bytes=1920106496, write_bytes=12288, read_chars=818344127, write_chars=373848710)
>>> p.io_counters()
pio(read_count=831200, write_count=5615929, read_bytes=1920892928, write_bytes=12288, read_chars=819155024, write_chars=373863380)
>>> p.io_counters()
pio(read_count=832971, write_count=5617618, read_bytes=1922465792, write_bytes=12288, read_chars=820778247, write_chars=373902937)
>>> 
>>> 
>>> #pause VLC
>>> #pause VLC
>>> p.io_counters()
pio(read_count=837762, write_count=5622914, read_bytes=1925013504, write_bytes=12288, read_chars=823227341, write_chars=374028455)
>>> 
>>> p.io_counters()
pio(read_count=837762, write_count=5623035, read_bytes=1925013504, write_bytes=12288, read_chars=823227341, write_chars=374037167)
>>> p.io_counters()
pio(read_count=837762, write_count=5623069, read_bytes=1925013504, write_bytes=12288, read_chars=823227341, write_chars=374039615)
>>> p.io_counters()
pio(read_count=837762, write_count=5623122, read_bytes=1925013504, write_bytes=12288, read_chars=823227341, write_chars=374043431)
```

So after pausing VLC, the counters do not go up too much anymore.

EDIT 2:

Overall disk IO:

```
>>> psutil.disk_io_counters()
sdiskio(read_count=870771, write_count=919577, read_bytes=43511045632, write_bytes=39517811712, read_time=1116828, write_time=2091068, read_merged_count=444494, write_merged_count=7546860, busy_time=1429112)
>>> psutil.disk_io_counters()
sdiskio(read_count=870777, write_count=919577, read_bytes=43511832064, write_bytes=39517811712, read_time=1116833, write_time=2091068, read_merged_count=444494, write_merged_count=7546860, busy_time=1429120)
```

**Reply by Safihre** [reply]
I'd suggest we keep it simpel, only detect if the process is in the list, not how much activity it shows.
Assuming that one closes VLC once done watching.

**Reply by thezoggy** [reply]
to play devils advocate here..

if you add logic for vlc being open, then what about if kodi is open. ok what if kodi is open but not playing anything. ok what about if plex is open but less than 3 streams.. etc etc. you can see where this becomes not fun to actually support and the context of an app being open vs doing something is not always apparent by just looking at a process list. 

this is why i dont think its wise for us to try and dictate, especially when most media apps can fire off a script before/after a movie starts. lots of time this is used so people can dim their smart lights / do x, can easily just extend that to send curl request to sab to pause/unpause. that way the media app is the trigger, rather than sab trying to guesstimate.

now, doing the inverse also shouldnt be hard with just a sab pre-queue script, where sab script sends a api/curl request to said app and logic there dictates if sab should pause for duration or whatever. little less elegant but could still accomplish the goal.

but honestly, most people just need to not download and play content off a spinner. if they used even a cheap ssd, generally they wont see any issues with sab activity. only people ive dealt with generally are those running sab on older hardware where they have one boot hdd (mechanical) where they are trying to do everything in the world on it. but again we already have various sab options to help out for that, which may help mask issues. also most media apps offer tuning to cache, so it can load more up in ram to deal with slower hdd/networks..  another thing that can be done to mask issues. i know back in the day (10 years ago when i ran sab on my htpc with everything else) i used to have to do that with kodi, increase its file cache so nothing i was doing on the box would affect it.

**Comment by opentoe** [comment]
Since  I use Plex Media Server to watch all my content it doesn't like to compete with other software. While SAB is downloading, extracting, or whatever and I'm trying to watch a high bitrate movie it these two programs do compete, even on a very powerful computer I have. What would it take to upgrade this thing to have SAB rocking and rolling and Plex throwing out 2160P at the same time with no issues?

Ryzen 5950x
64GB 4600 RAM
Asus Dark Hero Board
2x Western Digital Black 1T NVMe PCIe 4.0
Asus ROG Strix RTX-3090 video
and me

**Reply by Safihre** [reply]
Besides the feature requested here, on what OS are you? If it's on Linux you can use ionice and nice to give extracting and repairing super low system priority.

**Comment by sanderjo** [comment]
Interesting:

https://www.reddit.com/r/usenet/comments/w2p7x7/comment/igrdxrd/?utm_source=share&utm_medium=web2x&context=3

"Nzbget integrarion has a switch "nzbget_download" that sets the download state. A**nd plex has a entity "plex" that ready "x watching" you can setup an automation that triggers on the a change of plex entity "0 watching"** to enable nzbget_download and else disable nzbget_download."

So ... SABnzbd could pull that info from Plex ... ? I have no Plex, so I can't verify.
